src/artifact/decl.rs

```diff
@@ -0,0 +1,334 @@
+                    Decl::Import(ImportKind::Function) => Ok(()),
+                    _ => Err(ArtifactError::IncompatibleDeclaration {
+                        old: *self,
+                        new: other,
+                    }
+                    .into()),
+                }
+            }
+            // a previous data declaration can only be re-declared a data import, or it must match exactly the
+            // next declaration
+            decl @ Decl::Artifact(ADecl::Data { .. }) => match other {
+                Decl::Import(ImportKind::Data) => Ok(()),
+                other => {
+                    if decl == other {
+                        Ok(())
+                    } else {
+                        Err(ArtifactError::IncompatibleDeclaration {
+                            old: *self,
+                            new: other,
+                        }
+                        .into())
+                    }
+                }
+            },
+            // a previous function decl can only be re-declared a function import, or it must match exactly
+            // the next declaration
+            decl @ Decl::Artifact(ADecl::Function { .. }) => match other {
+                Decl::Import(ImportKind::Function) => Ok(()),
+                other => {
+                    if decl == other {
+                        Ok(())
+                    } else {
+                        Err(ArtifactError::IncompatibleDeclaration {
+                            old: *self,
+                            new: other,
+                        }
+                        .into())
+                    }
+                }
+            },
+            decl => {
+                if decl == other {
+                    Ok(())
+                } else {
+                    Err(ArtifactError::IncompatibleDeclaration {
+                        old: *self,
+                        new: other,
+                    }
+                    .into())
+                }
+            }
+        }
+    }
+    /// Is this an import (function or data) from a shared library?
+    pub fn is_import(&self) -> bool {
+        match *self {
+            Decl::Import(_) => true,
+            _ => false,
+        }
+    }
+    /// Is this a section?
+    pub fn is_section(&self) -> bool {
+        match *self {
+            Decl::Artifact(ADecl::DebugSection { .. }) => true,
+            _ => false,
+        }
+    }
+}
+
+pub struct FunctionImportDecl {}
+
+impl Default for FunctionImportDecl {
+    fn default() -> Self {
+        FunctionImportDecl {}
+    }
+}
+
+impl Into<Decl> for FunctionImportDecl {
+    fn into(self) -> Decl {
+        Decl::Import(ImportKind::Function)
+    }
+}
+
+pub struct DataImportDecl {}
+
+impl Default for DataImportDecl {
+    fn default() -> Self {
+        DataImportDecl {}
+    }
+}
+
+impl Into<Decl> for DataImportDecl {
+    fn into(self) -> Decl {
+        Decl::Import(ImportKind::Data)
+    }
+}
+
+pub struct FunctionDecl {
```

Comment:
All pub elements should have documentation

Reply:
Good point. I've been moving fast with this stuff. I really value the good docs this crate has and I'll make sure to add docs before this is ready.

Reply:
Added docs throughout